janus_collector/src/lib.rs

```diff
@@ -344,16 +345,39 @@ where
                 request.send().await
             },
         )
-        .await
-        .map_err(|res| match res {
-            Ok(response) => Error::Http(response.status()),
-            Err(error) => Error::HttpClient(error),
-        })?;
-
-        let status = response.status();
-        if status != StatusCode::SEE_OTHER {
-            return Err(Error::Http(status));
-        }
+        .await;
+
+        let response = match response_res {
+            // Successful response or unretryable error status code:
+            Ok(response) => {
+                let status = response.status();
+                if status == StatusCode::SEE_OTHER {
+                    response
+                } else if status.is_client_error() || status.is_server_error() {
+                    if let Ok(mut problem_details) = response.json::<HttpApiProblem>().await {
```

Comment:
This logic is slightly awkward as any failure to parse the body as a problem document will result in (a) the parsing error being silently ignored and (b) a problem document with only the status code populated being returned. My bigger concern is the former; silently dropping errors is not great.

This is complicated by the fact that (by my reading of the DAP spec) the body being a problem document is only a _suggestion_: an implementation is free to populate the body however it likes.

WDYT about the following logic:
1) If the response has content type `application/problem+json`, attempt to parse the body as a problem document. If parsing fails, log the error, and return a problem document based on the status code.
2) Otherwise, return a problem document based on the status code.

But I'm not sure if using the content type is mandatory; perhaps the DAP specification should be updated to say so? (if nothing else, problem documents can also be rendered as XML, and it'd be DAP-compliant to return an XML problem document as `application/problem+xml`; Janus wouldn't know how to handle this response.)